test/EResourceTest.cpp

```diff
@@ -0,0 +1,228 @@
+            }
+        }
+    }
+
+    GIVEN("The EResource is acquired shared")
+    {
+        kf::EResource resource;
+        bool acquired = resource.acquireShared();
+
+        THEN("The only shared lock is acquired")
+        {
+            REQUIRE(acquired == true);
+            REQUIRE(resource.isAcquiredExclusive() == false);
+            REQUIRE(resource.isAcquiredShared() > 0);
+        }
+
+        WHEN("The shared lock is released")
+        {
+            resource.release();
+
+            THEN("The resource is no longer locked")
+            {
+                REQUIRE(resource.isAcquiredExclusive() == false);
+                REQUIRE(resource.isAcquiredShared() == 0);
+            }
+        }
+    }
+
+    GIVEN("The EResource acquired exclusive")
+    {
+        kf::EResource resource;
+
+        WHEN("The resource is converted to shared")
+        {
+            bool exclusiveAcquired = resource.acquireExclusive();
+            REQUIRE(exclusiveAcquired == true);
+
+            resource.convertExclusiveToShared();
+
+            THEN("The lock is converted to shared")
+            {
+                REQUIRE(resource.isAcquiredExclusive() == false);
+                REQUIRE(resource.isAcquiredShared() > 0);
+            }
+
+            WHEN("The shared lock is released")
+            {
+                resource.release();
+
+                THEN("The resource is no longer locked")
+                {
+                    REQUIRE(resource.isAcquiredExclusive() == false);
+                    REQUIRE(resource.isAcquiredShared() == 0);
+                }
+            }
+        }
+    }
+
+    GIVEN("The EResource locked using the lock function")
+    {
+        kf::EResource resource;
+        resource.lock();
+
+        THEN("The resource is acquired exclusively and shared")
+        {
+            REQUIRE(resource.isAcquiredExclusive() == true);
+            REQUIRE(resource.isAcquiredShared() == 1);
+        }
+
+        WHEN("The resource is unlocked")
+        {
+            resource.unlock();
+
+            THEN("The resource is no longer locked")
+            {
+                REQUIRE(resource.isAcquiredExclusive() == false);
+                REQUIRE(resource.isAcquiredShared() == 0);
+            }
+        }
+    }
+
+    GIVEN("The EResource is locked shared using the lock_shared function")
+    {
+        kf::EResource resource;
+        resource.lock_shared();
+
+        THEN("The resource is acquired shared")
+        {
+            REQUIRE(resource.isAcquiredExclusive() == false);
+            REQUIRE(resource.isAcquiredShared() > 0);
+        }
+
+        WHEN("The resource is unlocked shared")
+        {
+            resource.unlock_shared();
+
+            THEN("The resource is no longer locked")
+            {
+                REQUIRE(resource.isAcquiredExclusive() == false);
+                REQUIRE(resource.isAcquiredShared() == 0);
+            }
+        }
+    }
+
+    GIVEN("The EResource and 2 threads")
```

Comment:
@SergiusTheBest I think it's time to introduce the kf-benchmark project, it is already added in my old branch, I can create a PR from it immediately